### ISSUE TITLE
new thumbnail matching

### ISSUE THREAD

**Issue by i30817** [libretro/RetroArch#16071]
I'd like to understand how the new thumbnail matching works, in the context of fetching the thumbnail to see how contributors should go about it @zoltanvb 

I understand that previously RA got its thumbnails from db_name of each entry to get the _system_ and the label to get the _name_ which then applies a denylist of illegal characters replacing them by underscore.

If the user had used the manual scanner without a .dat file, the label would be the same as the filename in all cases (instead of _just_ a vast majority due to no-intro \ redump naming schemes sometimes replacing illegal characters by legible alternatives instead of underscore)

Therefore, thumbnail makers tended to use the filename and in the few cases where it didn't match the .dat game name (label), added that variant too (with the illegal characters replaced by _)

Now, there was a commit a while ago to make the playlist present the filename, even in non manually scanned playlists. I thought that was simply cosmetic and it still used the underlying playlist label.

Then the new thumbnail download retry pr was committed, and I don't get it.

It apparently retries, first the filename thumbnail (why??? The previous label check _worked for both_ because the manual scanner places the filename on the label without a dat), then label (duplicating work in the majority of the cases of nonmatches) then the 'short' (cutting off the name in the first parenthesis, not sure if all variants of parenthesis or just the round ones).

In short, I think there might have been a misunderstsnding of the scheme the thumbnails and the manual\automatic scanner follow and there is a extra useless scanner call?

**Comment by i30817** [comment]
I also hope the recent minor changes commit uses 'the first db_name' just for databases and not thumbnails, because playlists actually save that by entry (for instance, for history playlist I think, or another hypothetical multiplatform playlist). If that is already so, ignore this post.

**Comment by zoltanvb** [comment]
Hi, thank you for caring about this topic. Not sure if I can put a response to all questions now, in this case I will return later.

About character replacements, [No-Intro naming convention](https://datomatic.no-intro.org/stuff/The%20Official%20No-Intro%20Convention%20(20071030).pdf) says:
 _the following Low ASCII characters are NOT allowed:
\ / : * ? " < > | `_
RetroArch code has one difference to this - it was like that and I did not make any change in this. Ampersand ("&") is also replaced to underscore. I checked the thumbnail repositories, and uploaded files do not have any of the other characters, but for the ampersand, there are sometimes double versions, as an example:
```
Itchy & Scratchy Game, The (USA, Europe).png
Itchy _ Scratchy Game, The (USA, Europe).png

```
There is no intended change for this part.

**Comment by i30817** [comment]
Yes I know about that replacement algorithm. Some dump sets replace & by ' and ' or vice versa, so I had to take care of that specifically in my fuzzy thumbnail downloader.

**Comment by zoltanvb** [comment]
_the label and the filename because of the way RetroArch can present both and both can be used as argument to get the thumbnail._
While this is technically true, it needed a specific setting from the user, as opposed to just working automatically.

_if you check that the png queried is not the same on all 3 calls_
This is definitely reasonable, and I had that in mind but the code structure turned out to be a bit more complex than what I hoped for. Anyways, now I have an idea on how to handle that, so it will come eventually.

_The vast majority of label\filename derived thumbs will be the same._
If it was just those databases where redump/no-intro sets are well formulated and maintained (where this is probably true), I would not have bothered with this. The main motivation was to handle those typical home computer systems better, where there is a huge e.g. TOSEC list with several different cracks, etc. of the same game.

The minor change commit enhances the case when there is no database associated, such as an entry in History playlist which was opened with Load Content directly. It should not have any effect in normal playlists, or when history entry is coming from a playlist.

**Comment by i30817** [comment]
> > the label and the filename because of the way RetroArch can present both and both can be used as argument to get the thumbnail.
> While this is technically true, it needed a specific setting from the user, as opposed to just working automatically.

The using the manual scanner is very common especially with sets that RetroArch won't scan. That by default always uses the filebame (if not providing a dat file). The automatic scanner always uses the label, since if it finds the game, it has a label in the database.

I suspect you're mentioning the _new_ setting to display playlist names as the filename, which I actually very much hoped didn't actually affect thumbnails download.

> The minor change commit enhances the case when there is no database associated, such as an entry in History playlist which was opened with Load Content directly.

Good to know.

**Comment by i30817** [comment]
Particularly I also think thumbnail makers are not very enthusiastic about adding TOSEC names, and I can't blame them myself. Even with my fuzzy thumbnail downloader, the nearest thing I have to a tosec set, the whdload set, which is a set without spaces between words, capitalized words, and that replaces and by &, I can only get a 75-80% coverage. Most of that is just missing thumbs in the server but not all I think.
